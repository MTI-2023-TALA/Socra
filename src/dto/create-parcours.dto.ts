import { IsDate, IsInt, IsNotEmpty, IsString, Max, Min } from 'class-validator';

export class CreateParcoursDto {
  @IsNotEmpty()
  @IsString()
  title!: string;

  @IsNotEmpty()
  @IsString()
  campus!: string;

  @IsNotEmpty()
  @IsInt()
  @Min(0)
  durationInMonths!: number;

  @IsNotEmpty()
  @IsString()
  type!: string;

  @IsNotEmpty()
  @IsInt()
  @Min(0)
  price!: number;

  @IsNotEmpty()
  @IsString()
  @Min(0)
  @Max(100)
  onSitePercentage!: number;

  @IsNotEmpty()
  @IsDate()
  beginDate!: Date;

  @IsNotEmpty()
  modules!: CreateModulesDto[];
}

class CreateModulesDto {
  @IsNotEmpty()
  @IsString()
  title!: string;

  @IsNotEmpty()
  @IsString()
  description!: string;
}
