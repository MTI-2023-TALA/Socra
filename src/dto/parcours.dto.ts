export class ParcoursDto {
  id!: string;
  title!: string;
  campus!: string;
  durationInMonths!: number;
  type!: string;
  price!: number;
  onSitePercentage!: number;
  beginDate!: Date;
  modules!: ModulesDto[];
  description!: string;
  createdAt!: Date;
}

class ModulesDto {
  title!: string;
  description!: string;
}
