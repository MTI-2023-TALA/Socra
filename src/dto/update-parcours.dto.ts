import { IsNotEmpty, IsString } from 'class-validator';

export class UpdateParcoursDto {
  @IsNotEmpty()
  @IsString()
  description!: string;
}
