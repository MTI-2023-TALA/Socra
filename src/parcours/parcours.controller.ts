import { Request, Response } from 'express';

import PDFDocument from 'pdfkit';
import { ParcoursControllerInterface } from './interfaces/parcours.controller.interface';
import { ParcoursServiceInterface } from './interfaces/parcours.service.interface';

export class ParcoursController implements ParcoursControllerInterface {
  constructor(private readonly parcoursService: ParcoursServiceInterface) {}

  public async getParcoursCheaperThan(req: Request, res: Response): Promise<void> {
    const parcours = await this.parcoursService.getParcoursCheaperThan(+req.params.price);
  }

  public async getParcoursWithCity(req: Request, res: Response): Promise<void> {
    const parcours = await this.parcoursService.getParcoursWithCity(req.params.city);
    res.send(parcours);
  }

  public async getAllParcours(req: Request, res: Response): Promise<void> {
    const parcours = await this.parcoursService.getAllParcours();
    res.send(parcours);
  }

  public async getParcoursById(req: Request, res: Response): Promise<void> {
    const parcours = await this.parcoursService.getParcoursById(req.params.id);
    if (!parcours) {
      res.status(404).send('Parcours not found');
      return;
    }
    res.send(parcours);
  }

  public async getParcoursByKeywords(req: Request, res: Response): Promise<void> {
    const parcours = await this.parcoursService.getParcoursByKeywords(req.body);
    res.send(parcours);
  }

  public async addParcours(req: Request, res: Response): Promise<void> {
    // TODO: verify what is sent in body and if attributes are correct
    const insertResult = await this.parcoursService.addParcours(req.body);
    res.send(insertResult);
  }

  public async updateParcours(req: Request, res: Response): Promise<void> {
    const id = req.params.id;
    const updatedParcours = await this.parcoursService.updateParcours(id, req.body);
    res.send(updatedParcours);
  }

  public async getParcoursPdf(req: Request, res: Response): Promise<void> {
    const doc = new PDFDocument({ bufferPages: true });

    const buffers: any[] = [];
    doc.on('data', buffers.push.bind(buffers));
    doc.on('end', () => {
      const pdfData = Buffer.concat(buffers);
      res
        .writeHead(200, {
          'Content-Length': Buffer.byteLength(pdfData as ArrayBuffer),
          'Content-Type': 'application/pdf',
          'Content-disposition': 'attachment;filename=test.pdf',
        })
        .end(pdfData);
    });
    await this.parcoursService.getParcoursPdf(req.params.id, doc);
  }
}
