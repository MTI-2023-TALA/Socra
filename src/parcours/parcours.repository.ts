import { Document, InsertOneResult, ObjectId, UpdateResult, WithId } from 'mongodb';

import { CreateParcoursDto } from '../dto/create-parcours.dto';
import { ParcoursRepositoryInterface } from './interfaces/parcours.repository.interface';
import { UpdateParcoursDto } from '../dto/update-parcours.dto';
import { parcoursCollection } from '../mongo';

export class ParcoursRepository implements ParcoursRepositoryInterface {
  public async getParcoursWithCity(city: string): Promise<WithId<Document>[]> {
    const regex = new RegExp(['^', city, '$'].join(''), 'i');
    return await parcoursCollection.find({ campus: regex }).toArray();
  }

  public async getAllParcours(): Promise<WithId<Document>[]> {
    return await parcoursCollection.find({}).sort({ createdAt: -1 }).toArray();
  }

  public async getParcoursById(id: string): Promise<WithId<Document> | null> {
    return parcoursCollection.findOne({ _id: new ObjectId(id) });
  }

  public async getParcoursByKeywords(keywords: string[]): Promise<WithId<Document>[]> {
    return parcoursCollection.find({ $text: { $search: keywords.join(' ') } }).toArray();
  }

  public async addParcours(createParcoursDto: CreateParcoursDto): Promise<InsertOneResult> {
    const createdAt = new Date(Date.now());
    return await parcoursCollection.insertOne({ ...createParcoursDto, createdAt });
  }

  public async updateParcours(id: string, updateParcoursDto: UpdateParcoursDto): Promise<UpdateResult> {
    return await parcoursCollection.updateOne(
      {
        _id: new ObjectId(id),
      },
      { $set: updateParcoursDto }
    );
  }
}
