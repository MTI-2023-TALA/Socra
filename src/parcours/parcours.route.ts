import { ParcoursController } from './parcours.controller';
import { app } from '../app';

export class ParcoursRoute {
  constructor(private readonly parcoursController: ParcoursController) {
    app.get('/parcours', async (req, res) => {
      await parcoursController.getAllParcours(req, res);
    });

    app.get('/parcours/:id', async (req, res) => {
      await parcoursController.getParcoursById(req, res);
    });

    app.get('/parcours/city/:city', async (req, res) => {
      await parcoursController.getParcoursWithCity(req, res);
    });

    app.get('/search', async (req, res) => {
      await parcoursController.getParcoursByKeywords(req, res);
    });

    app.post('/parcours', async (req, res) => {
      await parcoursController.addParcours(req, res);
    });

    app.patch('/parcours/:id', async (req, res) => {
      await parcoursController.updateParcours(req, res);
    });

    app.get('/parcours/pdf/:id', async (req, res) => {
      await parcoursController.getParcoursPdf(req, res);
    });
  }
}
