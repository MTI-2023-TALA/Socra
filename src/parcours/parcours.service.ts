import { Document, WithId } from 'mongodb';

import { CreateParcoursDto } from '../dto/create-parcours.dto';
import { ParcoursDto } from '../dto/parcours.dto';
import { ParcoursServiceInterface } from './interfaces/parcours.service.interface';
import { parcoursCollection } from '../mongo';

export class ParcoursService implements ParcoursServiceInterface {
  public async getAllParcours(): Promise<WithId<Document>[]> {
    return parcoursCollection.find({}).toArray();
  }

  public async addParcours(createParcoursDto: CreateParcoursDto): Promise<ParcoursDto> {
    const insertResult = await parcoursCollection.insertOne(createParcoursDto);
    return {
      id: insertResult.insertedId.toString(),
      ...createParcoursDto,
    };
  }
}
