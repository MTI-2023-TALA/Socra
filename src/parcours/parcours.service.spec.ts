import { Document, InsertOneResult, ObjectId, UpdateResult, WithId } from 'mongodb';

import { CreateParcoursDto } from '../dto/create-parcours.dto';
import { ParcoursRepositoryInterface } from './interfaces/parcours.repository.interface';
import { ParcoursService } from './parcours.service';
import { UpdateParcoursDto } from '../dto/update-parcours.dto';

class ParcoursRepositoryMock implements ParcoursRepositoryInterface {
  parcours: WithId<Document>[] = [
    {
      _id: new ObjectId('4edd40c86762e0fb12000003'),
      title: 'Génie Logiciel',
      createdAt: new Date('2020-01-01T00:00:00.000Z'),
      campus: 'Paris',
      durationInMonths: 24,
      type: 'Master',
      price: 12000,
      onSitePercentage: 50,
      beginDate: '2023-02-25',
      modules: [
        {
          title: `Systèmes d'information`,
          description: `Autour de la gestion d'information`,
        },
        {
          titre: 'ERP',
          description: 'Préparation Certification SAP',
        },
      ],
      description:
        'Un master dédié à l’étude des systèmes d’information ainsi que leur implémentation dans les entreprises',
    },
    {
      _id: new ObjectId('4edd40c86762e0fb12000004'),
      createdAt: new Date('2020-01-01T00:00:00.000Z'),
      title: 'Agronomie',
      campus: 'Strasbourg',
      durationInMonths: 24,
      type: 'DUT',
      price: 1000,
      onSitePercentage: 100,
      beginDate: '2023-01-01',
      modules: [
        {
          title: 'Trait des vaches',
          description: 'Apprentissage: Comment traire une vache comme un pro',
        },
        {
          titre: 'FOURCHE',
          description: 'Apprentissage: Combat à la fourche',
        },
      ],
      description: 'Un master dédié à devenir un bon agriculteur!',
    },
  ];

  async getAllParcours(): Promise<WithId<Document>[]> {
    return this.parcours;
  }

  async getParcoursById(id: string): Promise<WithId<Document> | null> {
    const p = this.parcours.find((parcours) => parcours._id.toString() === id);
    if (p === undefined) {
      return null;
    }
    return p;
  }

  async addParcours(createParcoursDto: CreateParcoursDto): Promise<InsertOneResult> {
    throw new Error('Method not implemented.');
  }

  async updateParcours(id: string, updateParcoursDto: UpdateParcoursDto): Promise<UpdateResult> {
    throw new Error('Method not implemented.');
  }
}

describe('ParcoursService', () => {
  let service: ParcoursService;
  let repository: ParcoursRepositoryMock;

  beforeEach(() => {
    repository = new ParcoursRepositoryMock();
    service = new ParcoursService(repository);
  });

  it('should be defined', () => {
    expect(repository).toBeDefined();
    expect(service).toBeDefined();
  });

  it('should call getAllParcours', async () => {
    const getAllParcours = jest.spyOn(repository, 'getAllParcours');

    const res = await service.getAllParcours();
    expect(res).toMatchSnapshot();
    expect(getAllParcours).toHaveBeenCalledTimes(1);
  });

  it('should call getParcoursById', async () => {
    const getParcoursById = jest.spyOn(repository, 'getParcoursById');

    const res = await service.getParcoursById('4edd40c86762e0fb12000003');
    expect(res).toMatchSnapshot();
    expect(getParcoursById).toHaveBeenCalledTimes(1);
  });

  it('should fail to retrieve one parcours by non-existing id', async () => {
    const getParcoursById = jest.spyOn(repository, 'getParcoursById');

    const res = await service.getParcoursById('4edd40c86762e0fb12000005');
    expect(res).toBeNull();
    expect(getParcoursById).toHaveBeenCalledTimes(1);
  });
});
