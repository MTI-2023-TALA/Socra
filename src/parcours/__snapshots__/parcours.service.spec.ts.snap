// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ParcoursService should be able to add new parcours 1`] = `
Object {
  "beginDate": 2023-03-05T00:00:00.000Z,
  "campus": "Toulouse",
  "description": "Master pour devenir Coach sportif ou athlète de haut-niveau",
  "durationInMonths": 24,
  "id": "4edd40c86762e0fb12000002",
  "modules": Array [
    Object {
      "description": "Cours de Badminton",
      "title": "Badminton",
    },
    Object {
      "description": "Cours de tennis",
      "title": "Tennis",
    },
    Object {
      "description": "Comment devenir un bon coach sportif",
      "title": "Coach Sportif 1'O'1",
    },
    Object {
      "description": "Bases du management",
      "title": "Management",
    },
  ],
  "onSitePercentage": 100,
  "price": 0,
  "title": "STAPS",
  "type": "Master",
}
`;

exports[`ParcoursService should call getAllParcours 1`] = `
Array [
  Object {
    "beginDate": 2023-02-25T00:00:00.000Z,
    "campus": "Paris",
    "createdAt": 2020-01-01T00:00:00.000Z,
    "description": "Un master dédié à l’étude des systèmes d’information ainsi que leur implémentation dans les entreprises",
    "durationInMonths": 24,
    "id": "4edd40c86762e0fb12000000",
    "modules": Array [
      Object {
        "description": "Autour de la gestion d'information",
        "title": "Systèmes d'information",
      },
      Object {
        "description": "Préparation Certification SAP",
        "titre": "ERP",
      },
    ],
    "onSitePercentage": 50,
    "price": 12000,
    "title": "Génie Logiciel",
    "type": "Master",
  },
  Object {
    "beginDate": 2023-01-01T00:00:00.000Z,
    "campus": "Strasbourg",
    "createdAt": 2020-01-01T00:00:00.000Z,
    "description": "Un master dédié à devenir un bon agriculteur!",
    "durationInMonths": 24,
    "id": "4edd40c86762e0fb12000001",
    "modules": Array [
      Object {
        "description": "Apprentissage: Comment traire une vache comme un pro",
        "title": "Trait des vaches",
      },
      Object {
        "description": "Apprentissage: Combat à la fourche",
        "titre": "FOURCHE",
      },
    ],
    "onSitePercentage": 100,
    "price": 1000,
    "title": "Agronomie",
    "type": "DUT",
  },
]
`;

exports[`ParcoursService should call getParcoursById 1`] = `null`;
